File: 437.路径总和III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // 双重递归 
    int s = 0;

    int pathSum(TreeNode* root, int sum) {
        if(!root)
            return 0;
        Sum(root, sum);
        pathSum(root->left, sum);
        pathSum(root->right, sum);
        return s;
    }
    
    void Sum(TreeNode* root, int sum){
        if(!root)
            return;
        // 这里要注意不能return，后面相加结果可能为0
        if(root->val == sum){
            ++s;
        }
        Sum(root->left, sum - root->val);
        Sum(root->right, sum - root->val);
    }
};

File: 树/437.路径总和III.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */
 var pathSum = function (root, targetSum) {
    let s = 0;

    const dfs =  (root) => {
        if (root === null)
            return;
        Sum(root, 0);
        dfs(root.left);
        dfs(root.right);
    }

    const Sum = (root, sum) => {
        if (root == null)
            return;
        sum += root.val;
        if (targetSum == sum) {
            ++s;
        }
        Sum(root.left, sum);
        Sum(root.right, sum);
    }

    dfs(root);
    return s;
};